### PR TITLE
add boot.initrd.unl0kr

### PR DESCRIPTION
**Pull request by tomfitzhenry**
###### Description of changes

[unl0kr](https://gitlab.com/cherrypicker/unl0kr) is a "Framebuffer-based disk unlocker for the initramfs based on LVGL".

Test via `nix-build -A nixosTests.systemd-initrd-luks-unl0kr`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/hardware/udev.nix [hunk 1]
```diff
@@ -10,7 +10,7 @@ let
 
   initrdUdevRules = pkgs.runCommand "initrd-udev-rules" {} ''
     mkdir -p $out/etc/udev/rules.d
-    for f in 60-cdrom_id 60-persistent-storage 75-net-description 80-drivers 80-net-setup-link; do
+    for f in 60-cdrom_id 60-input-id 60-persistent-storage 70-touchpad 75-net-description 80-drivers 80-net-setup-link; do
```
[the remark this answers is not in the document]
Comment (samueldr) [reply nixos/modules/services/hardware/udev.nix:13]: Needed by libinput [assuming libinput here too] to distinguish between device types

Ref: https://github.com/NixOS/mobile-nixos/commit/9ab16a6c6669eaa3652831ac9000495156b6a49a

File: nixos/modules/system/boot/unl0kr.nix [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    assertions = [
+      {
+        assertion = cfg.enable -> config.boot.initrd.systemd.enable;
+        message = "boot.initrd.unl0kr is only supported with boot.initrd.systemd.";
+      }
+    ];
+    boot.initrd.availableKernelModules = [
+      "evdev" # for entering pw
+      "bochs"
+    ];
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:28]: Are these hardware-specific?

File: nixos/modules/system/boot/unl0kr.nix [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    boot.initrd.systemd = {
+      storePaths = with pkgs; [
+        gnugrep
+        libinput
+        xkeyboard_config
+        systemd
+        unl0kr
+      ];
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:37]: These should probably be minimized, if possible. Like ideally we would be overriding `unl0kr` or whatever needs `systemd` to use `config.boot.initrd.systemd.package` instead. And only the files specifically needed from these other packages would be listed here.

File: nixos/modules/system/boot/unl0kr.nix [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      services = {
+        unl0kr-ask-password = {
+          description = "Forward Password Requests to unl0kr";
+          conflicts = [
+            "shutdown.target"
+            "emergency.service"
+          ];
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:44]: Probably also `initrd-switch-root.target`.

File: nixos/modules/system/boot/unl0kr.nix [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+            DIR=/run/systemd/ask-password/
+            # If a user has multiple encrypted disks, the requests might come in different times,
+            # so make sure to answer as many requests as we can. Once boot succeeds, other
+            # password agents will be responsible for watching for requests.
+            while [ -d $DIR ] && [ "$(ls -A $DIR/ask.*)" ];
+            do
+              for file in `ls $DIR/ask.*`; do
+                socket="$(cat "$file" | ${pkgs.gnugrep}/bin/grep "Socket=" | cut -d= -f2)"
+                ${pkgs.unl0kr}/bin/unl0kr | ${pkgs.systemd}/lib/systemd/systemd-reply-password 1 "$socket"
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:64]: Won't the request come with a prompt? Does unl0kr not have a way to print a prompt?

File: nixos/modules/system/boot/unl0kr.nix [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        };
+      };
+
+      paths = {
+        unl0kr-ask-password = {
+          description = "Forward Password Requests to unl0kr";
+          conflicts = [
+            "shutdown.target"
+            "emergency.service"
+          ];
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:77]: Probably also `initrd-switch-root.target`.

File: pkgs/tools/misc/unl0kr/default.nix [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+, libdrm
+, libinput
+, libxkbcommon
+, meson
+, ninja
+, pkg-config
+, scdoc
+, SDL2
+, SDL2_ttf
+}:
```
Comment (AndersonTorres) [inline pkgs/tools/misc/unl0kr/default.nix:18]:
```suggestion
}:

```

File: pkgs/tools/misc/unl0kr/default.nix [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+stdenv.mkDerivation rec {
```
Comment (AndersonTorres) [inline pkgs/tools/misc/unl0kr/default.nix:19]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/tools/misc/unl0kr/default.nix [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  pname = "unl0kr";
+  version = "2.0.0";
+
+  src = fetchFromGitLab {
+    domain = "gitlab.com";
+    owner = "cherrypicker";
+    repo = pname;
+    rev = version;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/unl0kr/default.nix:27]:
```suggestion
  repo = "unl0kr";
  rev = finalAttrs.version;
  fetchSubmodules = true;
```

File: pkgs/tools/misc/unl0kr/default.nix [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    hash = "sha256-KPP4Ol1GCAWqdQYlNtKQD/jx8A/xuHdvKjcocPMqWa0=";
+    fetchSubmodules = true;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/unl0kr/default.nix:29]:
```suggestion
```

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -25062,6 +25062,8 @@ with pkgs;
 
   unixODBCDrivers = recurseIntoAttrs (callPackages ../development/libraries/unixODBCDrivers { });
 
+  unl0kr = callPackage ../tools/misc/unl0kr { };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:25065]: Nowadays you can use `pkgs/by-name` structure. Just throw your default.nix file inside `pkgs/by-name/un/unl0kr/package.nix` and

```suggestion
```

**Review by AndersonTorres — DISMISSED** [review]

File: nixos/modules/system/boot/unl0kr.nix [hunk 12]
```diff
@@ -0,0 +1,91 @@
+{ config, lib, pkgs, utils, ... }:
```
Comment (samueldr) [inline nixos/modules/system/boot/unl0kr.nix:1]:
```suggestion
{ config, lib, pkgs, ... }:
```

`utils` is not used here.

File: nixos/modules/system/boot/unl0kr.nix [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+with lib;
+
```
Comment (samueldr) [inline nixos/modules/system/boot/unl0kr.nix:4]:
```suggestion
```

Minor nit, but having a `with` with such a wide scope should be avoided... It makes it harder down the line to reason where things come from, especially for newcomers. (Though less so when only a single `with` is at play...)

Whether you always use `lib.___`  or use `inherit (lib) ___` in the following `let` block is up to your preferences I guess.

**Review by samueldr — COMMENTED** [review]
Two nits, mainly, though I haven't tested, only read the changes to see if it looked fine.

File: nixos/modules/system/boot/unl0kr.nix (hunk 3 above)
Comment (ElvishJerricco) [reply nixos/modules/system/boot/unl0kr.nix]: @tomfitzhenry I still think these need to be reduced further. As is, you're including things like manpages. Like maybe just the `/bin` subdirectories or something like that.

**Comment by ElvishJerricco** [comment]
@ofborg test systemd-initrd-luks-unl0kr

**Review by ElvishJerricco — APPROVED** [review]
Assuming ofborg passes the test, LGTM

**Pull request merged**